crypto/cipher_extra/e_chacha20poly1305.c

```diff
@@ -346,3 +382,320 @@ const EVP_AEAD *EVP_aead_chacha20_poly1305(void) {
+        CRYPTO_poly1305_update(poly_ctx, out, inl);
+        cipher_ctx->len.text += inl;
+      } else {
+        // Decryption
+        // Update poly1305 with incoming ciphertext
+        CRYPTO_poly1305_update(poly_ctx, inp, inl);
+        cipher_chacha20_do_cipher(ctx, out, inp, inl);
+        cipher_ctx->len.text += inl;
+      }
+    }
+  }
+
+  // Process an |EVP_CipherFinal|
+  if (inp == NULL) {
+    uint8_t temp[POLY1305_TAG_LEN];
+    static const uint8_t padding[POLY1305_TAG_LEN] = {0};
+
+    // Finish AAD inp case there were no intermediate Update() calls
+    if (cipher_ctx->pad_aad) {
+      remainder = cipher_ctx->len.aad % POLY1305_TAG_LEN;
+      if (remainder != 0) {
+        CRYPTO_poly1305_update(poly_ctx, padding, sizeof(padding) - remainder);
+      }
+      cipher_ctx->pad_aad = 0;
+    }
+
+    // Apply padding for the text
+    remainder = cipher_ctx->len.text % POLY1305_TAG_LEN;
+    if (remainder != 0) {
+      CRYPTO_poly1305_update(poly_ctx, padding, sizeof(padding) - remainder);
+    }
+
+    // Apply len values
+    CRYPTO_poly1305_update(poly_ctx, (uint8_t *) &cipher_ctx->len,
+                           POLY1305_TAG_LEN);
+
+    // Compute the tag and write it to scratch or the cipher context
+    CRYPTO_poly1305_finish(poly_ctx, EVP_CIPHER_CTX_encrypting(ctx) ?
+      cipher_ctx->tag : temp);
+    cipher_ctx->poly_initialized = 0;
+
+    // Check the tags if we're decrypting
+    if (!EVP_CIPHER_CTX_encrypting(ctx)) {
+      if (CRYPTO_memcmp(temp, cipher_ctx->tag, cipher_ctx->tag_len)) {
+        return -1;
+      }
+    }
+  }
+  return (int32_t) inl;
+}
+
+static void cipher_chacha20_poly1305_cleanup(EVP_CIPHER_CTX *ctx) {
+  CIPHER_CHACHA_POLY_CTX *cipher_ctx = CCP_CTX(ctx);
+  if (cipher_ctx) {
+    OPENSSL_cleanse(ctx->cipher_data, sizeof(CIPHER_CHACHA_POLY_CTX));
+  }
+}
+
+// Returns a pointer to the first 64-byte alignment within |poly_ctx|.
+static void *cipher_chacha20_poly1305_aligned_poly(EVP_CIPHER_CTX *ctx) {
+  return (void *) (((uintptr_t) POLY_CTX(CCP_CTX(ctx)) + 63) & ~63);
+}
+
+static int32_t cipher_chacha20_poly1305_ctrl(EVP_CIPHER_CTX *ctx, int32_t type,
+     int32_t arg, void *ptr) {
+  CIPHER_CHACHA_POLY_CTX *cipher_ctx = CCP_CTX(ctx);
+  switch (type) {
+    case EVP_CTRL_INIT:
+      if (cipher_ctx == NULL) {
+        cipher_ctx = ctx->cipher_data = OPENSSL_malloc(ctx->cipher->ctx_size);
+      }
+      if (cipher_ctx == NULL) {
+        OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_INITIALIZATION_ERROR);
+        return 0;
+      }
+      cipher_ctx->len.aad = 0;
+      cipher_ctx->len.text = 0;
+      cipher_ctx->pad_aad = 0;
+      cipher_ctx->poly_initialized = 0;
+      cipher_ctx->tag_len = 0;
+      return 1;
+    case EVP_CTRL_COPY:
+      if (cipher_ctx && cipher_ctx->poly_initialized) {
+        // The poly1305 context needs to be aligned on a 64-byte boundary.
+        // The destination context doesn't necessarily have the same
+        // alignment so we have to fix that here.
```

Comment:
Will this be an issue if the user copies/moves the context? @justsmth what was the issue you hit, was it passing a XTS context by value instead which changed the alignment?

Reply:
This logic exists precisely to support copying the context. I set a flag in the cipher to force EVP to call this CTRL operation during all copy operations. This routine figures out the base pointer of the original poly-context and copies it to an aligned position in the destination context. It does costs us some redundant memory I/O though so I'd be interested in other ways to do this.